helix-term/src/ui/picker/handlers.rs

```diff
@@ -0,0 +1,168 @@
+                        return;
+                    };
+                    let Some(CachedPreview::Document(ref mut doc)) = picker.preview_cache.get_mut(&path) else {
+                        return
+                    };
+                    let diagnostics = helix_view::Editor::doc_diagnostics(
+                        &editor.language_servers,
+                        &editor.diagnostics,
+                        doc,
+                    );
+                    doc.replace_diagnostics(diagnostics, &[], None);
+                    doc.syntax = Some(syntax);
+                });
+            });
+        });
+    }
+}
+
+pub(super) struct DynamicQueryHandler<T: 'static + Send + Sync, D: 'static + Send + Sync> {
+    callback: Arc<DynQueryCallback<T, D>>,
+    last_query: Arc<str>,
+    query: Option<Arc<str>>,
+}
+
+impl<T: 'static + Send + Sync, D: 'static + Send + Sync> DynamicQueryHandler<T, D> {
+    pub(super) fn new(callback: DynQueryCallback<T, D>) -> Self {
+        Self {
+            callback: Arc::new(callback),
+            last_query: "".into(),
+            query: None,
+        }
+    }
+}
+
+impl<T: 'static + Send + Sync, D: 'static + Send + Sync> AsyncHook for DynamicQueryHandler<T, D> {
+    type Event = Arc<str>;
+
+    fn handle_event(&mut self, query: Self::Event, _timeout: Option<Instant>) -> Option<Instant> {
+        if query == self.last_query {
+            // If the search query reverts to the last one we requested, no need to
+            // make a new request.
+            self.query = None;
+            None
+        } else {
+            self.query = Some(query);
+            Some(Instant::now() + Duration::from_millis(275))
+        }
+    }
+
+    fn finish_debounce(&mut self) {
+        let Some(query) = self.query.take() else { return };
+        self.last_query = query.clone();
+        let callback = self.callback.clone();
+
+        crate::job::dispatch_blocking(move |editor, compositor| {
+            let Some(Overlay { content: picker, .. }) = compositor.find::<Overlay<Picker<T, D>>>() else {
+                return;
+            };
+            // Increment the version number to cancel any ongoing requests.
+            picker.version.fetch_add(1, atomic::Ordering::Relaxed);
+            picker.matcher.restart(false);
+            picker.dynamic_query_running = true;
+            let injector = picker.injector();
+            let get_options = (callback)(&query, editor, picker.editor_data.clone(), &injector);
+            tokio::spawn(async move {
+                if let Err(err) = get_options.await {
+                    log::info!("Dynamic request failed: {err}");
+                }
+
+                crate::job::dispatch(|_editor, compositor| {
+                    let Some(Overlay { content: picker, .. }) = compositor.find::<Overlay<Picker<T, D>>>() else {
+                        return;
+                    };
+                    picker.dynamic_query_running = false;
+                }).await;
```

Comment:
I think its possible that this causes race conditions (so dynamic_query_running is false when it shouldn't be). It also dispatches to the main loop when it really doens't need to.

Instead we can try the following: `injector` is refcounted. We can ask nucleo what its injector refcount is. If its > 1 then the dynamic query is running (which also happens to cover things like the file picker wich is not a dyn picker so its a more general solution). I will look into adding an API for that to nucleo

Reply:
Sounds good :+1:

I believe we talked about this briefly in the past and thought we might need to expose the refcount from nucleo